skimage/segmentation/slic_superpixels.py

```diff
@@ -122,10 +237,36 @@ def slic(image, n_segments=100, compactness=10., max_iter=10, sigma=0,
         # Add channel as single last dimension
         image = image[..., np.newaxis]
 
+    if multichannel and (convert2lab or convert2lab is None):
+        if image.shape[-1] != 3 and convert2lab:
+            raise ValueError("Lab colorspace conversion requires a RGB image.")
+        elif image.shape[-1] == 3:
+            image = rgb2lab(image)
+
+    # initialize cluster centroids for desired number of segments
+    update_centroids = False
+    if use_mask:
+        mask = np.asarray(mask, dtype=np.bool)
+        if mask.ndim == 2:
+            mask = np.ascontiguousarray(mask[np.newaxis, ...])
+        if mask.shape != image.shape[:3]:
+            raise ValueError("image and mask should have the same shape.")
+        # Step 1 of the algorithm [3]_
+        centroids, steps = _get_mask_centroids(mask, n_segments, spacing)
+        update_centroids = True
+        start_label = 1
+    else:
+        centroids, steps = _get_grid_centroids(image, n_segments)
+        mask = np.ones((0, 1, 1), dtype=np.bool)
+        start_label = 0
+        warnings.warn("Labels' indexing start from 0. " +
```

Comment:
I don't think this should be the way to handle deprecations.

A user should be able to write code on the day the warning is issued that will do two things:
1. Silence the warning.
2. Ensure their code stays correct when the scikit-image version gets upgraded.

Reply:
Hello @hmaarrfk, I am not sure to understand your comment here.
1. to silence the warning, the user may use `warnings.catch_warnings`
2. the warning message is made for that! The user is informed and he is responsible of making his code correct when skimage gets upgraded. "We are all consenting adults" :-D
Now, the warning message has to be as clear as possible and it is may be actually not. It basically misses the version number when the change will be made.
We can also think about exposing the `start_label` as a parameter of the function set to `0` by default (to keep the actual behavior) and change it to `1` when we decide to address #3722 
What do you think?

Reply:
My opinion is close to what you think about the spirit of warnings:
1. Receive a warning in 0.16, silence the warning and if you mind about it, take an action that make your code valid anyway.
2. That action is valid in 0.17 and the warning won't appear
3. No future action is necessary in 0.18 to fix the code.

:smile: 

By the way, it is somehow strange to introduce an additional parameter, but it's fine.

Reply:
@rfezzani,

> to silence the warning, the user may use warnings.catch_warnings

This is a very verbose way to handle warnings, and most Python users won't be familiar with this. It's particularly cumbersome to use in interactive settings. A keyword argument is much easier for people to adapt to.

> "We are all consenting adults"

Even adults get annoyed at having to do extra work. =) I think @hmaarrfk's principles are exactly correct: warnings should result in a changed function call that still does what the user wants and that will work "forever". We haven't kept to this in the past but I really think it's the best user experience.

Reply:
:sweat_smile: I still believe that retro-compatibility is most of the time a lot of effort impacting a very small portion of user.
Code versioning is made to handle that, and people needing such stability usually fix software versions.
Now I understand your point of view, It is really a matter of choice.